examples/wandb_sweeps/sweep.py

```diff
@@ -41,10 +42,19 @@ def run_sweep(argv: List[str]) -> None:
     We encapsulate each run by using a separate subprocess, which ought to
     ensure that memory is returned (etc.).
     """
-    try:
-        subprocess.check_call(argv)
-    except subprocess.CalledProcessError as error:
-        logging.error("Subprocess error: %s", error)
+    process = subprocess.Popen(argv, stderr=subprocess.PIPE, text=True)
+    # subprocess.Popen is nonblocking, so the while loop waits for
```

Comment:
So, creating a `Popen` instance doesn't block, but looping over it in the idiomatic way I recommend does.

I wasn't completely positive I was right about this, so I made a toy example. The subprocess waits 10 seconds then prints a line to stderr. The wrapper calls that subprocess using `Popen`. I think you'll see it works just fine without this additional complexity (see instructions there):

https://gist.github.com/kylebgorman/191c4496f5d0c3b98c025b631b470b2c

Were this not blocking, the loop in the wrapper would have terminated long before the 10 seconds are up. 

Of course it may be the case that there is some reason my toy example is non-representative of what you're trying to do here, and if so let me know. But please take a look...everything else looks fine.

Reply:
@kylebgorman Thanks for the gist! I've extended slow.py in the gist to return a code, but it seems the return code isn’t being captured. I suspect that this is due to a gap between when the pipe closes and when the process actually ends and returns a code.

With subprocess.Popen, it seems we need either subprocess.poll() or subprocess.communicate() to ensure the subprocess is properly terminated. While subprocess.communicate() feels more idiomatic, my sample runs show that it doesn’t consistently capture the stack trace. If we decide to use subprocess.communicate(), we might consider using subprocess.run() instead for simplicity.

Additionally, I was able to simplify the while loop to make it more concise and idiomatic. Between these approaches, which would you recommend? Thanks again! =)

Reply:
That's not a problem either; you just call `process.wait()` at the end of the loop, which returns the exit code. I updated the example to make sure I wasn't crazy. `while True:` and `.readline` are both code smells, so I recommend at least trying to adapt this design.